### PR TITLE
add function to calculate Buchsbaum frequency to formulary.frequencies.py

### PR DESCRIPTION
**Pull request by svincena**
add function to calculate Buchsbaum frequency to formulary.frequencies.py

## Motivation and context

The function is useful for determining the Buchsbaum (aka ion-ion hybrid) frequency in a plasma with two ion species.
This is the location of a parallel cutoff and perpendicular resonance for Alfvénic wave propagation. The function is a little inconvenient to calculate from scratch as it depends on the local magnetic field and the ratio of densities between the two ion species. It can be used as a diagnostic for laboratory and magnetospheric measurements of this density ratio.  It is important for fast wave propagation in tokamak plasmas and establishes the existence of trapped modes in planetary magnetospheres.

## Related issues

??None??

### CODE REVIEW

File: plasmapy/formulary/frequencies.py [hunk 1]
```diff
@@ -516,3 +518,114 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+    ion1 : `~plasmapy.particles.particle_class.Particle`
+        Representation of ion species #1 (e.g., 'p' for protons, 'D+'
+        for deuterium, or 'He-4 +1' for singly ionized helium-4). If no
+        charge state information is provided, then species #1 is assumed
+        to be singly charged.
+    ion2 : `~plasmapy.particles.particle_class.Particle`
+        Representation of ion species #2 (e.g., 'p' for protons, 'D+'
+        for deuterium, or 'He-4 +1' for singly ionized helium-4). If no
+        charge state information is provided, then species #2 is assumed
+        to be singly charged.
```
Comment (StanczakDominik) [inline plasmapy/formulary/frequencies.py:561]: Can we, and should we, do a "same as above for the second particle"?

File: plasmapy/formulary/frequencies.py [hunk 2]
```diff
@@ -516,3 +518,114 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+    <Quantity 0.50168706>
+
+    """
+
+    omega_c1 = gyrofrequency(B, ion1, signed=False, Z=Z1)
+    omega_c2 = gyrofrequency(B, ion2, signed=False, Z=Z2)
+    omega_p1 = plasma_frequency(n1, ion1, z_mean=Z1)
+    omega_p2 = plasma_frequency(n2, ion2, z_mean=Z2)
+
+    return np.sqrt((omega_p1**2 * omega_c2**2 + omega_p2**2 * omega_c1**2) / (omega_p1**2 + omega_p2**2))
```
Comment (StanczakDominik) [inline plasmapy/formulary/frequencies.py:621]: Since only the squares are used, precalculate them above!

**Review by StanczakDominik — CHANGES_REQUESTED** [review]
Looks good! Got a few minor suggestions first.

File: plasmapy/formulary/frequencies.py (hunk 1 above)
Comment (svincena) [reply plasmapy/formulary/frequencies.py]: Shortened description for ion2

File: plasmapy/formulary/frequencies.py [hunk 3]
```diff
@@ -516,3 +518,114 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+    }
+)
+@angular_freq_to_hz
+def Buchsbaum_frequency(
+    B: u.T,
+    n1: u.m**-3,
+    n2: u.m**-3,
+    ion1: ParticleLike,
+    ion2: ParticleLike,
+    Z1=None,
```
[the remark this answers is not in the document]
Comment (svincena) [reply plasmapy/formulary/frequencies.py]: I think I did it right?!

[the remark this answers is not in the document]
Comment (StanczakDominik) [reply plasmapy/formulary/frequencies.py]: Yup! :)

File: plasmapy/formulary/tests/test_frequencies.py [hunk 4]
```diff
@@ -187,3 +194,19 @@ def test_upper_hybrid_frequency():
...
+def test_Buchsbaum_frequency():
+    r"""Test the Buchsbaum_frequency function in frequencies.py."""
+
+    with pytest.raises(InvalidParticleError):
+       Buchsbaum_frequency(1.0 * u.T, 5e19 * u.m**-3, 5e19 * u.m**-3,
+       "norwegian jarlsberg","proton")
+
+    with pytest.raises(InvalidParticleError):
+       Buchsbaum_frequency(1.0 * u.T, 5e19 * u.m**-3, 5e19 * u.m**-3,
+       "proton", "venezuelan beaver cheese")
```
Comment (StanczakDominik) [inline plasmapy/formulary/tests/test_frequencies.py:207]: Nice to see my favorite sketch! 😁

File: plasmapy/formulary/frequencies.py [hunk 5]
```diff
@@ -516,3 +519,112 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+
+
+wbb_ = Buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.Buchsbaum_frequency`."""
+
+wii_ = Buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.Buchsbaum_frequency`."""
+
+wbi_ = Buchsbaum_frequency
+"""Alias to `~plasmapy.formulary.frequencies.Buchsbaum_frequency`."""
```
Comment (StanczakDominik) [inline plasmapy/formulary/frequencies.py:630]: I've never been a fan of those aliases, but leaving that aside - do you think it's a good idea to have three aliases point to this same function? Smells like an antipattern.

File: plasmapy/formulary/frequencies.py [hunk 6]
```diff
@@ -516,3 +519,112 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+    B: u.T,
+    n1: u.m**-3,
+    n2: u.m**-3,
+    ion1: ParticleLike,
+    ion2: ParticleLike,
+    Z1: Optional[float] = None,
+    Z2: Optional[float] = None
+) -> u.rad / u.s:
+    r"""
+    Return the Buchsbaum frequency for a two-ion-species plasma.
```
Comment (StanczakDominik) [inline plasmapy/formulary/frequencies.py:541]: My earlier point on this still stands - it's still a restatement of the function name + list of arguments. I'd still prefer
```suggestion
    Return the ion-ion hybrid frequency for a two-ion-species plasma.
```
but of course, between the two of us you're the domain expert, and if this you say this is good enough, guess it is. Just making sure the earlier point is acknowledged and not accidentally missed!

File: changelog/1828.feature.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Added a function to return the Buchsbaum frequency to plasmapy.formulary.frequencies
```
Comment (namurphy) [inline changelog/1828.feature.rst:1]:
```suggestion
Added the `~plasmapy.formulary.frequencies.Buchsbaum_frequency` function.
```

File: plasmapy/formulary/frequencies.py [hunk 8]
```diff
@@ -516,3 +519,101 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+        Representation of ion species #1 (e.g., 'p' for protons, 'D+'
+        for deuterium, or 'He-4 +1' for singly ionized helium-4). If no
+        charge state information is provided, then species #1 is assumed
+        to be singly charged.
+    ion2 : `~plasmapy.particles.particle_class.Particle`
+        Representation of ion species #2 (same behavior as for ion1).
+    Z1 : `float` or `~astropy.units.Quantity`, optional
+        The charge state for ion species #1. If not provided, it defaults to the charge number of ``ion1``.
+    Z2 : `float` or `~astropy.units.Quantity`, optional
+        The charge state for ion species #2. If not provided, it defaults to the charge number of ``ion2``.
```
Comment (namurphy) [inline plasmapy/formulary/frequencies.py:561]:
```suggestion
    B : `~astropy.units.Quantity`
        The magnetic field magnitude in units convertible to tesla.

    n1 : `~astropy.units.Quantity`
        Particle number density of ion species #1 in units convertible to m\ :sup:`-3`.

    n2 : `~astropy.units.Quantity`
        Particle number density of ion species #2 in units convertible to m\ :sup:`-3`.

    ion1 : `~plasmapy.particles.particle_class.Particle`
        Representation of ion species #1 (e.g., 'p' for protons, 'D+'
        for deuterium, or 'He-4 +1' for singly ionized helium-4). If no
        charge state information is provided, then species #1 is assumed
        to be singly charged.

    ion2 : `~plasmapy.particles.particle_class.Particle`
        Representation of ion species #2 (same behavior as for ion1).

    Z1 : `float` or `~astropy.units.Quantity`, optional
        The charge state for ion species #1. If not provided, it 
        defaults to the charge number of ``ion1``.

    Z2 : `float` or `~astropy.units.Quantity`, optional
        The charge state for ion species #2. If not provided, it
        defaults to the charge number of ``ion2``.
```
...slight changes to formatting.

File: plasmapy/formulary/frequencies.py [hunk 9]
```diff
@@ -516,3 +519,101 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+        The Buchsbaum frequency of the plasma in units of radians per second.
+        Setting keyword ``to_hz=True`` will apply the factor of :math:`1/2π`
+        and yield a value in Hz.
+
+    Raises
+    ------
+    `TypeError`
+        If the magnetic field is not a `~astropy.units.Quantity` or
+        ``particle`` is not of an appropriate type.
+    `ValueError`
```
Comment (namurphy) [inline plasmapy/formulary/frequencies.py:575]:
```suggestion
    `ValueError`

```

File: plasmapy/formulary/frequencies.py [hunk 10]
```diff
@@ -516,3 +519,101 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+        If units are not provided, SI units are assumed.
+
+    Notes
+    -----
+    In a magnetized plasma, the presence of two ion species allows the
+    perpendicular component of the cold-plasma dielectric coefficient
+    :math:`\epsilon_{\perp}` to vanish at an angular frequency referred
+    to as the Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called
+    the bi-ion hybrid resonance frequency :cite:p: `thompson:1995`, or ion-ion hybrid frequency
+    :cite:p:`vincena:2013`.  This frequency can be defined as:
```
Comment (namurphy) [inline plasmapy/formulary/frequencies.py:591]:
```suggestion
    the bi-ion hybrid resonance frequency :cite:p: `thompson:1995`, or
    ion-ion hybrid frequency :cite:p:`vincena:2013`.  This frequency
    can be defined as:
```
...to keep the docstring ≲72 characters.

File: plasmapy/formulary/frequencies.py [hunk 11]
```diff
@@ -516,3 +519,101 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+
+    .. math::
+        \omega_{BB} \equiv \sqrt{\frac{\omega_{p1}^{2}\omega_{c2}^{2}
+            + \omega_{p2}^{2}\omega_{c1}^{2}}{\omega_{p2}^{2}+\omega_{p2}^{2}}}
```
Comment (namurphy) [inline plasmapy/formulary/frequencies.py:595]:
```suggestion
        ω_{BB} ≡ \sqrt{\frac{ω_{p1}^2 ω_{c2}^2
            + ω_{p2}^2 ω_{c1}^2}{ω_{p2}^2 + ω_{p2}^2}}
```
Might make the math a bit more readable when using `help()`.

File: plasmapy/formulary/tests/test_frequencies.py [hunk 12]
```diff
@@ -32,6 +33,9 @@
         (wc_, gyrofrequency),
         (wlh_, lower_hybrid_frequency),
         (wuh_, upper_hybrid_frequency),
+        (wbb_, Buchsbaum_frequency),
+        (wii_, Buchsbaum_frequency),
+        (wbi_, Buchsbaum_frequency),
```
Comment (namurphy) [inline plasmapy/formulary/tests/test_frequencies.py:38]:
```suggestion
```
It looks like the aliases were deleted, in which case we can remove the tests too.

File: plasmapy/formulary/tests/test_frequencies.py [hunk 13]
```diff
@@ -187,3 +191,19 @@ def test_upper_hybrid_frequency():
...
+       "norwegian jarlsberg","proton")
+
+    with pytest.raises(InvalidParticleError):
+       Buchsbaum_frequency(1.0 * u.T, 5e19 * u.m**-3, 5e19 * u.m**-3,
+       "proton", "venezuelan beaver cheese")
+
+    assert np.isclose(Buchsbaum_frequency(B=0.1*u.T, n1=1e18*u.m**-3, n2=1e18*u.m**-3,
+        ion1="proton", ion2="He-4 +1").value, 4805575.93140432)
+#
+#    assert_can_handle_nparray(Buchsbaum_frequency)
```
Comment (namurphy) [inline plasmapy/formulary/tests/test_frequencies.py:209]:
```suggestion

```
...unless you were still planning on adding this test.

File: plasmapy/formulary/tests/test_frequencies.py [hunk 14]
```diff
@@ -187,3 +191,19 @@ def test_upper_hybrid_frequency():
...
+    with pytest.raises(InvalidParticleError):
+       Buchsbaum_frequency(1.0 * u.T, 5e19 * u.m**-3, 5e19 * u.m**-3,
+       "norwegian jarlsberg","proton")
+
+    with pytest.raises(InvalidParticleError):
+       Buchsbaum_frequency(1.0 * u.T, 5e19 * u.m**-3, 5e19 * u.m**-3,
+       "proton", "venezuelan beaver cheese")
+
+    assert np.isclose(Buchsbaum_frequency(B=0.1*u.T, n1=1e18*u.m**-3, n2=1e18*u.m**-3,
+        ion1="proton", ion2="He-4 +1").value, 4805575.93140432)
```
Comment (namurphy) [inline plasmapy/formulary/tests/test_frequencies.py:207]:
```suggestion
    assert u.isclose(Buchsbaum_frequency(B=0.1*u.T, n1=1e18*u.m**-3, n2=1e18*u.m**-3,
        ion1="proton", ion2="He-4 +1").value, 4805575.93140432 * u.Hz)
```
Would it help to use `astropy.units.isclose` here?

File: docs/bibliography.bib [hunk 15]
```diff
@@ -288,6 +299,29 @@ @article{stringer:1963
...
+pages = {625-634},
+year = {1995},
+issn = {0032-0633},
+doi = {https://doi.org/10.1016/0032-0633(94)00197-Y},
+url = {https://www.sciencedirect.com/science/article/pii/003206339400197Y},
+author = {P. Thompson and M.K. Dougherty and D.J. Southwood},
+}
+@article{vincena:2013,
+author = {Vincena,S. T.  and Farmer,W. A.  and Maggs,J. E.  and Morales,G. J. },
+title = {Investigation of an ion-ion hybrid Alfv\'{e}n wave resonator},
```
Comment (namurphy) [inline docs/bibliography.bib:316]:
```suggestion
title = {Investigation of an ion-ion hybrid Alfvén wave resonator},
```

File: docs/bibliography.bib [hunk 16]
```diff
@@ -75,6 +75,17 @@ @article{braginskii:1965
    volume = 1,
    pages = 205
 }
+@article{buchsbaum:1960,
+author = {Buchsbaum,S. J. },
```
Comment (namurphy) [inline docs/bibliography.bib:79]:
```suggestion
author = {Buchsbaum, S. J. },
```

File: docs/bibliography.bib (hunk 15 above)
Comment (svincena) [reply docs/bibliography.bib:316]: It's always been my habit to escape accents, but if this is resolved with UTF, then so be it! :)

File: plasmapy/formulary/tests/test_frequencies.py (hunk 14 above)
Comment (svincena) [reply plasmapy/formulary/tests/test_frequencies.py:207]: I just copied examples above my function in frequencies.py, so I have no idea. I'll commit your suggestion

File: plasmapy/formulary/tests/test_frequencies.py (hunk 13 above)
Comment (svincena) [reply plasmapy/formulary/tests/test_frequencies.py]: I copied examples from those above in 'frequencies.py' however, it failed on this line, so I commented it out to get it to pass. I think it should be able to do np arrays, but I don't know why it failed.

File: plasmapy/formulary/tests/test_frequencies.py (hunk 12 above)
Comment (svincena) [reply plasmapy/formulary/tests/test_frequencies.py:38]: Indeed.

File: plasmapy/formulary/frequencies.py (hunk 11 above)
Comment (svincena) [reply plasmapy/formulary/frequencies.py:595]: Again, I don't speak UTF, so this is foreign to me, but so let it be written, so let it be done!

File: plasmapy/formulary/frequencies.py (hunk 6 above)
Comment (svincena) [reply plasmapy/formulary/frequencies.py:541]: I see this definition as functionally the same as 'gyrofrequency.'
I'd welcome someone else to have a look at the wording, but I think it's ok.

File: plasmapy/formulary/frequencies.py (hunk 5 above)
Comment (svincena) [reply plasmapy/formulary/frequencies.py]: This is not a fundamental quantity. I removed the aliases. Thanks.

File: plasmapy/formulary/tests/test_frequencies.py (hunk 4 above)
Comment (svincena) [reply plasmapy/formulary/tests/test_frequencies.py:201]: Thanks!

**Review by StanczakDominik — APPROVED** [review]

**Comment by namurphy** [comment]
Some of the problems with the doc build seem to be associated with the machinery we put in `plasmapy.formulary.parameters` when we deprecated it.  Since we gave several releases worth of notice and said we were going to remove it in 0.9.0, I went ahead and submitted #1833 to remove `parameters.py` completely.

**Comment by StanczakDominik** [comment]
Just so you don't feel like you're going insane: the docstring line length limits were just merged today.

**Comment by StanczakDominik** [comment]
I'll also merge main in to include Nick's parameters removal.

**Comment by StanczakDominik** [comment]
Merged in main to hopefully fix doc issues

File: plasmapy/formulary/frequencies.py [hunk 17]
```diff
@@ -516,3 +519,108 @@ def upper_hybrid_frequency(B: u.T, n_e: u.m**-3) -> u.rad / u.s:
...
+    : `~astropy.units.UnitsWarning`
+        If units are not provided, SI units are assumed.
+
+    Notes
+    -----
+    In a magnetized plasma, the presence of two ion species allows the
+    perpendicular component of the cold-plasma dielectric coefficient
+    :math:`\epsilon_{\perp}` to vanish at an angular frequency referred
+    to as the Buchsbaum frequency :cite:p:`buchsbaum:1960`, also called
+    the bi-ion hybrid resonance frequency :cite:p: `thompson:1995`, or
```
Comment (namurphy) [inline plasmapy/formulary/frequencies.py:599]:
```suggestion
    the bi-ion hybrid resonance frequency :cite:p:`thompson:1995`, or
```

**Comment by StanczakDominik** [comment]
pre-commit.ci autofix

File: tox.ini [hunk 18]
```diff
@@ -287,22 +286,17 @@ enable-extensions =
...
-    plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:FS003,W505
+    plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py:FS003
+    plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:FS003
     plasmapy/formulary/__init__.py:E402,F403
     plasmapy/formulary/collisions/__init__.py:E402,F403
     plasmapy/formulary/collisions/lengths.py:SIM114
     # The decorators.py ignore is temporary pending #1057
     plasmapy/particles/decorators.py:SIM102
     plasmapy/particles/ionization_state_collection.py:SIM102
     # Files with long lines in docstrings
```
Comment (namurphy) [inline tox.ini:302]: Remove this line too?

**Comment by StanczakDominik** [comment]
pre-commit.ci autofix

**Comment by StanczakDominik** [comment]
I've just realized one thing:

```
formulary/frequencies.rst:1: WARNING: autosummary: stub file not found 'plasmapy.formulary.frequencies.plasmapy.formulary.frequencies.Buchsbaum_frequency'. Check your autosummary_generate setting.
```

Note the `plasmapy.formulary.frequencies` appears **twice**. It's some silly circular dependency or whatever thing.

**Comment by StanczakDominik** [comment]
pre-commit.ci blocked on #1860

**Comment by namurphy** [comment]
Thank you for adding this, @svincena!  I remember that this got delayed since we were getting a weird (intermittent?) test failure, but that seems to have gone away.  I'm happy we'll have this in `v2023.5.0`!

**Pull request merged**